File: src/AetherNet.Security/PreKeyExchange/PreKeyExchangeService.cs
// SPDX-License-Identifier: MIT

using System.Collections.Concurrent;
using System.Text.Json;
using AetherNet.Constants;
using AetherNet.Protocol;
using AetherNet.Routing;
using AetherNet.Security.Models;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;

namespace AetherNet.PreKeys;

/// <summary>
/// Default mesh pre-key exchange service. Directed request/response — never broadcast — so bundle
/// requests do not leak identity-interest to the whole mesh. Transport only: the host wires the
/// published bundle in (<see cref="SetLocalBundle"/>) and consumes received bundles out
/// (<see cref="BundleReceived"/>) via ISignalProtocolService.
/// </summary>
public sealed class PreKeyExchangeService : IPreKeyExchangeService
{
    private static readonly JsonSerializerOptions JsonOptions = new();

    private readonly IMeshSender _sender;
    private readonly ILogger<PreKeyExchangeService> _logger;

    private PreKeyBundle? _local;
    private readonly ConcurrentDictionary<string, PreKeyBundle> _received = new(StringComparer.Ordinal);

    public event EventHandler<PreKeyBundleReceivedEventArgs>? BundleReceived;

    public PreKeyExchangeService(IMeshSender sender, ILogger<PreKeyExchangeService>? logger = null)
    {
        _sender = sender ?? throw new ArgumentNullException(nameof(sender));
        _logger = logger ?? NullLogger<PreKeyExchangeService>.Instance;
    }

    /// <inheritdoc />
    public void SetLocalBundle(PreKeyBundle bundle)
        => _local = bundle ?? throw new ArgumentNullException(nameof(bundle));

    /// <inheritdoc />
    public PreKeyBundle? GetLocalBundle() => _local;

    /// <inheritdoc />
    public async Task<Guid> RequestBundleAsync(string peerUhid, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(peerUhid);

        var requestId = Guid.NewGuid();
        var payload = new PreKeyRequestPayload { RequestId = requestId, RequesterUhid = _sender.LocalUhid };
        var packet = new MeshPacket
        {
            Type = PacketType.PreKeyRequest,
            SourceUhid = _sender.LocalUhid,
            DestinationUhid = peerUhid,
            Ttl = ProtocolConstants.DefaultTtl,
            Payload = JsonSerializer.SerializeToUtf8Bytes(payload, JsonOptions),
        };

        var delivered = await _sender.SendAsync(packet, peerUhid, cancellationToken).ConfigureAwait(false);
        _logger.LogDebug("PreKey request {Req} → {Peer} delivered={Delivered}", requestId, peerUhid, delivered);
        return requestId;
    }

    /// <inheritdoc />
    public Task<bool> HandleAsync(MeshPacket packet, CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(packet);
        return packet.Type switch
        {
            PacketType.PreKeyRequest => HandleRequestAsync(packet, cancellationToken),
            PacketType.PreKeyResponse => Task.FromResult(HandleResponse(packet)),
            _ => Task.FromResult(false),
        };
    }

    private async Task<bool> HandleRequestAsync(MeshPacket packet, CancellationToken cancellationToken)
    {
        PreKeyRequestPayload? body;
        try
        {
            body = JsonSerializer.Deserialize<PreKeyRequestPayload>(packet.Payload, JsonOptions);
        }
        catch (JsonException ex)
        {
            _logger.LogDebug(ex, "PreKeyRequest from {Source}: malformed payload — dropped", packet.SourceUhid);
            return false;
        }
        if (body is null)
            return false;

        var local = _local;
        if (local is null)
        {
            _logger.LogDebug("PreKeyRequest {Req} from {Source}: no local bundle set — ignored", body.RequestId, packet.SourceUhid);
            return false;
        }

        var replyTo = !string.IsNullOrEmpty(body.RequesterUhid) ? body.RequesterUhid : packet.SourceUhid;
        var payload = PreKeyResponsePayload.FromBundle(body.RequestId, local);
        var reply = new MeshPacket
        {
            Type = PacketType.PreKeyResponse,
            SourceUhid = _sender.LocalUhid,
            DestinationUhid = replyTo,
            Ttl = ProtocolConstants.DefaultTtl,
            Payload = JsonSerializer.SerializeToUtf8Bytes(payload, JsonOptions),
        };

        var delivered = await _sender.SendAsync(reply, replyTo, cancellationToken).ConfigureAwait(false);
        _logger.LogDebug("PreKey response {Req} → {Peer} delivered={Delivered}", body.RequestId, replyTo, delivered);
        return true;
    }

    private bool HandleResponse(MeshPacket packet)
    {
        PreKeyResponsePayload? body;
        try
        {
            body = JsonSerializer.Deserialize<PreKeyResponsePayload>(packet.Payload, JsonOptions);
        }
        catch (JsonException ex)
        {
            _logger.LogDebug(ex, "PreKeyResponse from {Source}: malformed payload — dropped", packet.SourceUhid);
            return false;
        }
        if (body is null || string.IsNullOrEmpty(body.Uhid))
            return false;

        var bundle = body.ToBundle();
        _received[body.Uhid] = bundle;
        BundleReceived?.Invoke(this, new PreKeyBundleReceivedEventArgs
        {
            RequestId = body.RequestId,
            FromUhid = packet.SourceUhid,
            Bundle = bundle,
        });
        return true;
    }

    /// <inheritdoc />
    public PreKeyBundle? GetReceivedBundle(string uhid)
        => _received.TryGetValue(uhid, out var b) ? b : null;
}

File: kotlin/src/main/kotlin/aethernet/prekey/PreKeyExchangeService.kt
// SPDX-License-Identifier: MIT

package aethernet.prekey

import aethernet.AetherNetConstants
import aethernet.protocol.MeshPacket
import aethernet.protocol.PacketType
import aethernet.routing.MeshSender
import aethernet.security.PreKeyBundle
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

/**
 * Mesh pre-key exchange over [PacketType.PreKeyRequest] (25) and [PacketType.PreKeyResponse] (26).
 * Closes the "how does a peer get another peer's [PreKeyBundle] over the mesh" gap the messaging layer
 * previously left out-of-band.
 *
 * A node publishes its current bundle via [setLocalBundle] (the host produces it with
 * [aethernet.security.SignalProtocol.generatePreKeyBundle]). A peer asks for it with [requestBundle];
 * the responder replies with its bundle; the requester caches it and fires [onBundleReceived]. This
 * service is the mesh TRANSPORT of bundles — the host performs the actual X3DH by feeding the received
 * bundle to [aethernet.security.SignalProtocol.processPreKeyBundle] (Signal-canonical: no key agreement
 * happens here).
 *
 * Directed request/response — never broadcast — so bundle requests do not leak identity-interest to
 * the whole mesh. Mirrors the C# `PreKeyExchangeService` and the Kotlin
 * [aethernet.videocall.VideoCallControlService] (directed [MeshSender.send], nullable-lambda event).
 */
class PreKeyExchangeService(
    private val sender: MeshSender
) {
    /**
     * Raised when a peer's pre-key bundle arrives in a [PacketType.PreKeyResponse]. Nullable-lambda
     * event mechanism — matches the other Kotlin services and C# PreKeyExchangeService.BundleReceived.
     */
    var onBundleReceived: ((PreKeyBundleReceived) -> Unit)? = null

    private var local: PreKeyBundle? = null
    private val received = ConcurrentHashMap<String, PreKeyBundle>()

    /** Set (or replace) this node's published bundle — served in reply to inbound requests. */
    fun setLocalBundle(bundle: PreKeyBundle) {
        local = bundle
    }

    /** The currently-published local bundle, or null if none has been set. */
    fun getLocalBundle(): PreKeyBundle? = local

    /** The most recently received bundle for [uhid], or null. */
    fun getReceivedBundle(uhid: String): PreKeyBundle? = received[uhid]

    /**
     * Ask [peerUhid] for its pre-key bundle: mint a request id and send a directed
     * [PacketType.PreKeyRequest]. Returns the new request id (echoed by the response).
     */
    suspend fun requestBundle(peerUhid: String): UUID {
        require(peerUhid.isNotEmpty()) { "peerUhid must not be empty" }

        val requestId = UUID.randomUUID()
        val payload = PreKeyRequestPayload(requestId = requestId, requesterUhid = sender.localUhid)
        val packet = MeshPacket(
            type = PacketType.PreKeyRequest,
            sourceUhid = sender.localUhid,
            destinationUhid = peerUhid,
            ttl = AetherNetConstants.DEFAULT_TTL,
            payload = payload.toJsonBytes()
        )

        sender.send(packet, peerUhid)
        return requestId
    }

    /**
     * Process an incoming pre-key packet. On [PacketType.PreKeyRequest], reply with the local bundle
     * (if set). On [PacketType.PreKeyResponse], cache the peer bundle and fire [onBundleReceived].
     * Returns false for the wrong packet type, a malformed payload, or a request received when no
     * local bundle is set.
     */
    suspend fun handle(packet: MeshPacket): Boolean = when (packet.type) {
        PacketType.PreKeyRequest -> handleRequest(packet)
        PacketType.PreKeyResponse -> handleResponse(packet)
        else -> false
    }

    private suspend fun handleRequest(packet: MeshPacket): Boolean {
        val json = packet.payload.toString(Charsets.UTF_8)
        val body = PreKeyRequestPayload.fromJson(json) ?: return false

        val bundle = local ?: return false

        val replyTo = if (body.requesterUhid.isNotEmpty()) body.requesterUhid else packet.sourceUhid
        val payload = PreKeyResponsePayload.fromBundle(body.requestId, bundle)
        val reply = MeshPacket(
            type = PacketType.PreKeyResponse,
            sourceUhid = sender.localUhid,
            destinationUhid = replyTo,
            ttl = AetherNetConstants.DEFAULT_TTL,
            payload = payload.toJsonBytes()
        )

        sender.send(reply, replyTo)
        return true
    }

    private fun handleResponse(packet: MeshPacket): Boolean {
        val json = packet.payload.toString(Charsets.UTF_8)
        val body = PreKeyResponsePayload.fromJson(json) ?: return false

        val bundle = body.toBundle()
        received[body.uhid] = bundle
        onBundleReceived?.invoke(
            PreKeyBundleReceived(
                requestId = body.requestId,
                fromUhid = packet.sourceUhid,
                bundle = bundle
            )
        )
        return true
    }
}
